### ISSUE TITLE
Socket closed on the server side, indefinitely allows writing on the client side

### ISSUE THREAD

**Issue by fedemennite** [guillep/pharo-core#15614]
**Bug description**
Socket closed on the server side, indefinitely allows writing on the client side

**To Reproduce**
Evalute:
```
	| listener clientSocket serverSocket clientStream serverStream writeResult connectedResult |
	listener := Socket newTCP.
	[listener listenOn: 0 backlogSize: 4.

	clientSocket := Socket newTCP.
	clientSocket connectTo: #[127 0 0 1] port: listener localPort.
	clientSocket waitForConnectionFor: 1.
	self assert: clientSocket isConnected.

	serverSocket := listener waitForAcceptFor: 1.
	self assert: serverSocket isConnected.
	] ensure:[listener destroy].

	clientStream := SocketStream on: clientSocket.
	serverStream := SocketStream on: serverSocket.
	
	
	serverStream close.
	writeResult  := [clientStream nextPutAll:'A line of text'; flush.
	true] on: ConnectionClosed do:[:ex | false].  
	connectedResult := clientStream isConnected.
	
	clientStream ifNotNil:[clientStream destroy].
	serverStream ifNotNil:[serverStream destroy].

	writeResult -> connectedResult
```

**Expected behavior**
The evaluation above should result in false -> false meaning that:
a write operation on a disconnected socket should fail, possibly with a ConnectionClosed errors or as a bare minimum #isConnected should return false after the first write operation failure.

**Version information:**
 - OS: at least on Windows and Linux
 - Version: [e.g. 1.6.1]
 - Pharo Version 10

**Expected development cost**
A few hours for somebody who knows the area.
I would says it's more concerning the impact for existing code.

**Comment by daniels220** [comment]
There are a few things going on here:

Look at the first line of `SocketStream>>flush`: `((outNextToWrite > 1) and: [socket isOtherEndClosed not]) ifTrue: [...]`—and there is no `ifFalse:`. In other words, if the socket is closed, `SocketStream` simply ignores the `flush` request! At first glance I would agree this is surprising behavior, but it might also be by design that an error is only ever signaled _once_ when a connection fails, and subsequent write attempts are silently swallowed. However that error _isn't_ actually signaled properly—see below.

Now, in your example, this is never hit, since you only call `flush` once, so that's not the problem. You might wonder, though, why you never see an error. In your case, the _first_ reason is that when you attempt the send, the socket still thinks it's connected (you cannot assume that the `serverSocket closeAndDestroy` synchronously affects the state of the client socket—in general, assume that all operations that hit the network take however long would be the least convenient for you 😁). And if you follow "Browse definitions" from `SocketStream>>flush`, you'll see that it ultimately reduces to a loop calling `Socket>>sendSomeData:count:startIndex:for:`, which _first_ does a `waitForSendDoneFor:`, then calls the primitive that actually does the send. This means that the _last_ chunk of data sent does _not_ wait for the send to complete.

So in your example, since the amount of data you send _easily_ fits in any conceivable buffer size, it all gets sent at once and the method returns without checking. If you send some more data and flush again—`2 timesRepeat: [clientStream nextPutAll:'A line of text'; flush].`—or send a large amount of data (the default `SO_SNDBUF` is 128kiB, I haven't tested whether that's the actual threshold), you'll find that `connectedResult` at least is now false. There should be an error signaled in here too though—see below as to why not. And once it _has_ been signaled, you could keep trying to send _more_ data and `SocketStream` will happily ignore you.

This I'm more ambivalent about—I think there may be good performance reasons not to block the application on small sends, but it is surprising that small sends can silently fail like this. However it's worth noting that with several small sends in a row, the subsequent ones _will_ block, so you can still have performance issues—if you really want to minimize waits, you use a stream, as in your example, and only flush it when enough data has accumulated to fill the buffer. If performance is your primary concern, it's still good for the `flush` to return immediately so you can start accumulating data to fill the next buffer, but if you're relying on all the data actually being sent when `flush` returns, that's not the case.

For now, you can manually `waitForSendDoneFor:` yourself (though there is no passthrough API on `SocketStream`—this could be added). It might also make sense to either add another method that does a flush with a final wait, or make it so that calling `flush` with an empty buffer still does a wait—effectively making `flush` behave like `fsync`, where if you _really_ want things synced dammit, you have to call it twice in a row. I don't know whether to consider that elegant or ugly 😅.

But why, as mentioned above, do you not get an error if you send-and-flush again immediately? The SocketStream _doesn't_ bail on you, that's not the problem—that would only happen if you waited between the two, I think. No, now the problem is that the error handling code in `SocketStream>>flush` is broken—it says `shouldSignal ifFalse: ["swallow"]`, but without a corresponding `ifTrue:` block, this is a useless statement, and the error is swallowed no matter what. If I add `ifTrue: [ex pass]` here, the error is correctly passed through, and you'd _think_ we'd end up with `writeResult = false`.

But no! Instead we get a debugger on an uncaught `ConnectionTimedOut`! Thing is, the abovementioned `waitForSendDoneFor:` makes no distinction between a timeout and the connection being closed. The read equivalent is `waitForDataFor:ifClosed:ifTimedOut:`—it would definitely make sense to refactor `waitForSendDoneFor:` to work this way as well so that the correct _kind_ of error can be signaled.

TL;DR:
1. Fix error handling in `SocketStream>>flush`
2. Distinguish timeout/close in `waitForSendDone` family
3. Consider passthrough API on `SocketStream` for `waitForSendDoneFor:` for applications that need to guarantee writes make it to the network (though in practice, this is a pretty weak guarantee—the only way to guarantee _receipt_, in my experience, is an application-level acknowledgement from the other end).

**Comment by daniels220** [comment]
To be clear, I've done some work on this and I'm hoping to get a PR together...not sure how soon, maybe in a couple steps as the most blatant issues are easy to fix.

**Comment by fedemennite** [comment]
Thanks for the fix and the detailed review of the issue!

**Comment by daniels220** [comment]
I should add for posterity: I now understand part of why `waitForSendDoneFor:` is such a weak guarantee—it's actually misnamed, in that what it waits for is not "all buffered data has been sent", but only "there is space in the send buffer", i.e. _some_ buffered data has been sent so we can "top off" the buffer if we want. It would be _possible_ to wait for "send buffer is empty" (which if no error occurs means an ACK has been received for all previously-buffered data), but would require new plugin code to accomplish, and from what I've read—I'm hardly an expert on sockets programming—this would _still_ not provide the same level of guarantee as an _application_-level acknowledgement from the other side.
